File: Data Structures & Algorithms/products-of-array-discluding-self/submission-2.cpp
class Solution {
public:
    vector<int> productExceptSelf(vector<int>& nums) {
        int prod = 1;
        int z_cnt = 0;

        for(int i = 0; i < nums.size(); i++){
            if (nums[i] == 0){
                z_cnt += 1;
            } else {
                prod *= nums[i];
            }
        }

        vector<int> res(nums.size(), 0);
        if(z_cnt > 1){
            return res;
        }

        for(int j = 0; j < nums.size(); j++){
            if (z_cnt == 1){
                if (nums[j] == 0){
                    res[j] = prod;
                }
            } else {
                res[j] = prod / nums[j];
            }
        }
        return res;
    }
};

File: Data Structures & Algorithms/permutations/submission-2.py
class Solution:
    def permute(self, nums: List[int]) -> List[List[int]]:
        res = []
        seen = []

        for k in range(len(nums)):
            seen.append(False)

        def backtrack(i, path, length, seen):
            if (length == len(nums)):
                res.append(path[:])
                return
            
            for j in range(len(nums)):
                if i < len(nums) and (not seen[j]):
                    path.append(nums[j])
                    seen[j] = True
                    backtrack(i+1, path, length + 1, seen)
                    path.pop()
                    seen[j] = False
                    backtrack(i+1, path, length, seen)

                # if i < len(nums) and nums[j] not in path:
                #     path.append(nums[j])
                #     backtrack(i + 1, path, length + 1)
                #     path.pop()
                #     backtrack(i + 1, path, length)

        backtrack(0, [], 0, seen)
             

        return res
        # res = []

        # for i in nums:
        #     temp = []
        #     temp.append(i)
        #     for j in nums:
        #         if j not in temp:
        #             temp.append(j)
        #         for k in nums:
        #             if k not in temp:
        #                 temp.append(k)
        #             if temp not in res:
        #                 res.append(temp)
        #                 while len(temp) > 1:
        #                     temp.pop()
        #                 continue
        
        # return res

        